hidi/factorization.py

```diff
@@ -0,0 +1,202 @@
+import warnings
+import numpy as np
+import collections
+from numpy.random import permutation
+from sklearn.decomposition import TruncatedSVD
+from hidi.transform import Transform
+
+# Catch annoying warnings from nimfa
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    import nimfa as nf
+
+
+class W2VStringTransform(Transform):
+    """
+    Takes a pandas Dataframe and transforms it into a
+    string
+
+    :param n_shuffles: The number of suffles for the
+    `item_id`.
+    :type n_shuffles: int
+    """
+    def __init__(self, n_shuffles=3, **w2v_kwargs):
+        self.w2v_kwargs = w2v_kwargs
+        self.n_shuffles = n_shuffles
+
+    def transform(self, df_temp, **kwargs):
+        """
+        :param df: a pandas Dataframe with two columns:
+        :code:`link_id`, :code:`item_id`
+        :type df: pandas.Dataframe
+        :rtype: a string of shuffled item_id
+        """
+        df = df_temp
+        if ('link_id' in df.index) | ('item_id' in df.index):
+            df.reset_index(inplace=True)
+
+        df.set_index('link_id', inplace=True)
+        words = ''
+        for index in df.index.unique():
+            temp_item_id_list = df.loc[index].item_id
+            item_id_list = temp_item_id_list
+            for i in range(self.n_shuffles - 1):
+                item_id_list = np.append(item_id_list,
+                                         permutation(temp_item_id_list))
+            joined_shuffled_list = ' '.join(str(x) for x in item_id_list)
+            words = ' '.join([words, joined_shuffled_list]).strip()
+        return words, kwargs
```

Comment:
At some point, we should see how long this takes, and benchmark it against `StringIO`. I have no idea which would be faster, but I think both should be possible.